### PR TITLE
CAS-4 Feature/Switch user to Sample app

### LINKED ISSUE #196: Add switch user to sample app
This will make it easier to test disconnect/connect behavior

### PR DESCRIPTION
**Pull request by adrian09h**
# Submit a pull request

close #196

### CODE REVIEW

**Comment by elevenetc** [comment]
@adrian09h lets add bottom navigation and new profile page. 
And we will need to update also how developers can configure user ids and other information. We probably need to have at least 3 users, each one with photo, name, id and token. So to avoid having huge config in  `build.gradle` we can use config json file and load it. Might look about like this:
```groovy
def usersConfig = loadJson(" custom user config from .gitignore ")
if( usersConfig == null ) usersConfig = loadJson( " default usef config ")
buildConfigField "String", "USERS_CONFIG", usersConfig
```
And parse them in some dedicated place:
```
class UsersStorage {
  init {
    val config = parseJson(UserConfig::class, BuildConfig.USERS_CONFIG)
  }

  fun getCurrentUser() {
    return config[0]
  }

  fun isLoggedIn() {
    return sharedPreferences.get("logged-in", condig.isLoggedIn)
  }
}
```

**Comment by elevenetc** [comment]
@adrian09h Trying to test the branch, but getting connection error with all users:
```java
D/StreamWebSocketService: WebSocket # 1 Response : {"error":{"code":43,"message":"JWTAuth error: signature is not valid. Make sure the token is created using the secret for API key \"d2q3juekvgsf\"","StatusCode":401,"duration":""}}
D/StreamWebSocketService: WebSocket # 1 Closing : 1000 / 
E/ChannelListViewModel: onError for loading the channels JWTAuth error: signature is not valid. Make sure the token is created using the secret for API key "d2q3juekvgsf"
E/ChannelListFragment: Failed to establish websocket connection. Code 43 message JWTAuth error: signature is not valid. Make sure the token is created using the secret for API key "d2q3juekvgsf"
```

**Comment by adrian09h** [comment]
> @adrian09h Trying to test the branch, but getting connection error with all users:
> 
> ```java
> D/StreamWebSocketService: WebSocket # 1 Response : {"error":{"code":43,"message":"JWTAuth error: signature is not valid. Make sure the token is created using the secret for API key \"d2q3juekvgsf\"","StatusCode":401,"duration":""}}
> D/StreamWebSocketService: WebSocket # 1 Closing : 1000 / 
> E/ChannelListViewModel: onError for loading the channels JWTAuth error: signature is not valid. Make sure the token is created using the secret for API key "d2q3juekvgsf"
> E/ChannelListFragment: Failed to establish websocket connection. Code 43 message JWTAuth error: signature is not valid. Make sure the token is created using the secret for API key "d2q3juekvgsf"
> ```

Hmmm, we are using `qk4nn7rpcn75` in sample app, have no idea about API key "d2q3juekvgsf"
Anyhow I have resolved conflicts and updated  minor stuff.(worked well without the issue.)
@elevenetc Can you please check it again?

File: build.gradle [hunk 1]
```diff
@@ -6,6 +6,28 @@ buildscript {
         buildToolsVersion = "28.0.3"
         minSdkVersion = 21
         compileSdkVersion = 28
+        appCompatVersion = '1.1.0'
```
Comment (elevenetc) [inline build.gradle]: That's nice.

File: build.gradle [hunk 2]
```diff
@@ -17,13 +39,12 @@ buildscript {
...
-//        classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.8.1'
-        classpath 'com.github.dcendents:android-maven-gradle-plugin:2.1'
-        classpath 'com.google.gms:google-services:4.3.2'
-        classpath 'io.fabric.tools:gradle:1.31.1'  // Crashlytics plugin
-        classpath "de.mannodermaus.gradle.plugins:android-junit5:1.5.1.0" // jUnit5 plugin
-        classpath 'com.dicedmelon.gradle:jacoco-android:0.1.4'
+        classpath "com.android.tools.build:gradle:$gradleVersion"
+        classpath "com.github.dcendents:android-maven-gradle-plugin:$gradlePluginVersion"
+        classpath "com.google.gms:google-services:$googleServiceVersion"
+        classpath "io.fabric.tools:gradle:$fabricVersion"  // Crashlytics plugin
```
Comment (elevenetc) [inline build.gradle]: `Crashlytics plugin` that's too obvious.

File: build.gradle [hunk 3]
```diff
@@ -17,13 +39,12 @@ buildscript {
...
+        classpath "de.mannodermaus.gradle.plugins:android-junit5:$mannodermausVersion" // jUnit5 plugin
```
Comment (elevenetc) [inline build.gradle]: `jUnit5 plugin` that's also too obvious.

File: sample/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -14,20 +14,22 @@
...
             <intent-filter>
                 <action android:name="OPEN_ACTIVITY_1" />
+
                 <category android:name="android.intent.category.DEFAULT" />
             </intent-filter>
         </activity>
-        <activity android:name=".ChannelActivity"  android:parentActivityName=".MainActivity">
-
-        </activity>
+        <activity android:name=".MainActivity"></activity>
```
Comment (elevenetc) [inline sample/src/main/AndroidManifest.xml]: As we introduced other activities lets rename `MainActivity` to `HomeActivity`.

File: sample/src/main/assets/users-config.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
+[
```
Comment (elevenetc) [inline sample/src/main/assets/users-config.json]: This json also should have apiKey. Currently it's provided in `build.gradle`, but should be moved here to keep all users information in a single place. So structure might be like this:
```json
{
  apps: [
    {
      apiKey:"",
      users: [{id:""}, {id:""}]
    }
  ]
}
```

File: sample/src/main/java/io/getstream/chat/example/LoginActivity.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+import io.getstream.chat.example.adapter.UserListItemAdapter;
+import io.getstream.chat.example.utils.UserStorage;
+
+public class LoginActivity extends AppCompatActivity {
+
+    private static final String TAG = LoginActivity.class.getSimpleName();
+
+    private UserListItemAdapter adapter;
+    private ListView lv_users;
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/LoginActivity.java]: `ListView` is seems an oldish choice. Let's switch to `RecyclerView`.

File: sample/src/main/java/io/getstream/chat/example/LoginActivity.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_login);
+
+        lv_users = findViewById(R.id.lv_users);
+        UserStorage.init(this);
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/LoginActivity.java]: Lets move it to application. `LoginActivity` shouldn't care about initialisation of other parts of the app.

File: sample/src/main/java/io/getstream/chat/example/LoginActivity.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import android.view.View;
+import android.widget.AdapterView;
+import android.widget.ListView;
+import android.widget.Toast;
+
+
+import io.getstream.chat.example.adapter.UserListItemAdapter;
+import io.getstream.chat.example.utils.UserStorage;
+
+public class LoginActivity extends AppCompatActivity {
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/LoginActivity.java]: There should be also header with subtitle from `MainActivity`. 
It's convenient to know which SDK version we're using from login screen.

File: sample/src/main/java/io/getstream/chat/example/utils/UserStorage.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import androidx.annotation.Nullable;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.util.List;
+
+public class UserStorage {
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/utils/UserStorage.java]: There is no need to make such kind of classes with all static methods. It's ok for utilities where all static methods are independent and eventually might be tested independently.

But this one defines interconnected methods with common business logic, so it should be hidden behind interface that we can replace and test it:
```java
interface UserStorage {
  UserConfig getCurrentUser();
  //... other methods
}
```
```java
class StreamUserStorage implements UserStorage {
  public static UserStorage get() { return instance; }
  @Override UserConfig getCurrentUser() {}
} 
```

File: sample/src/main/java/io/getstream/chat/example/utils/UserStorage.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    private static List<UserConfig>users;
+
+    private static final String USER_DATA_FILENAME = "users-config.json";
+    private static final String CURRENT_USER_ID = "current-user-id";
+    private static final String APP_PREFERENCE = "AppPreferences";
+
+    public static void init(Context mContext) {
+        context = mContext;
+        preferences = context.getSharedPreferences(APP_PREFERENCE, Activity.MODE_PRIVATE);
+        parseJson();
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/utils/UserStorage.java]: We should also get list of users from `BuildConfig`. 
That's important to let developers define and use their own users and apiKeys.

File: sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        client.disconnect();
+
+        Intent intent = new Intent(getContext(), LoginActivity.class);
+        startActivity(intent);
+        getActivity().finish();
+    }
+
+
+    @Override
+    public void onAttach(Context context) {
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java]: Unused override.

File: sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        super.onAttach(context);
+    }
+
+    @Override
+    public void onDetach() {
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java]: Unused override.

File: sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        // Inflate the layout for this fragment
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java]: Too obvious comment.

File: sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+import io.getstream.chat.example.LoginActivity;
+import io.getstream.chat.example.databinding.FragmentProfileBinding;
+import io.getstream.chat.example.utils.UserStorage;
+
+public class ProfileFragment extends Fragment {
+
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
```
Comment (elevenetc) [inline sample/src/main/java/io/getstream/chat/example/view/fragment/ProfileFragment.java]: Unused override.

**Pull request merged**